onnxruntime/core/framework/memory_info.h

```diff
@@ -0,0 +1,216 @@
+#pragma once
+#include "core/graph/basic_types.h"
+#include "core/framework/alloc_kind.h"
+#include "core/framework/data_types.h"
+#include "core/framework/execution_plan_base.h"
+#include "core/framework/mem_pattern.h"
+#include "core/framework/tensor.h"
+#include "core/framework/ort_value_name_idx_map.h"
+
+#include <iomanip>
+
+namespace onnxruntime {
+using IntervalT = std::pair<size_t, size_t>;
+using OrtValueIndex = int;
+using OrtValueName = std::string;
+
+struct MemoryInfoPerTensor {
+  MemoryBlock planned_block;
+  MemoryBlock alloced_block;
+};
+
+struct MemoryInfoMap {
+ public:
+  MemoryInfoMap() = default;
+
+  inline void AddPlannedMemory(const OrtValueIndex& idx, const MemoryBlock& mb) {
+    map_[idx].planned_block = mb;
+  }
+
+  inline void AddAllocMemory(const OrtValueIndex& idx, MemoryBlock& mb) {
+    map_[idx].alloced_block = mb;
+    if (ptr_offset == 0 || (ptr_offset > mb.offset_ && mb.offset_ != 0)) {
+      ptr_offset = mb.offset_;
+    }
+  }
+
+  inline const MemoryBlock& GetPlannedMemory(const OrtValueIndex& idx) const {
+    ORT_ENFORCE(map_.find(idx) != map_.end());
+    return map_.at(idx).planned_block;
+  }
+
+  inline size_t GetPlannedAddress(const OrtValueIndex& idx) const {
+    ORT_ENFORCE(map_.find(idx) != map_.end());
+    return map_.at(idx).planned_block.offset_;
+  }
+
+  inline size_t GetPlannedSize(const OrtValueIndex& idx) const {
+    ORT_ENFORCE(map_.find(idx) != map_.end());
+    return map_.at(idx).planned_block.size_;
+  }
+  size_t GetAllocAddress(const OrtValueIndex& idx, bool raw = false) const {
+    ORT_ENFORCE(map_.find(idx) != map_.end());
+    if (raw) {
+      return map_.at(idx).alloced_block.offset_;
+    } else {
+      return map_.at(idx).alloced_block.offset_ - ptr_offset;
+    }
+  }
+
+  inline size_t GetAllocSize(const OrtValueIndex& idx) const {
+    ORT_ENFORCE(map_.find(idx) != map_.end());
+    return map_.at(idx).alloced_block.size_;
+  }
+
+  inline bool Contain(const OrtValueIndex& idx) {
+    return map_.find(idx) != map_.end();
+  }
+
+  inline void clear() {
+    map_.clear();
+  }
+
+  std::unordered_map<onnxruntime::OrtValueIndex, onnxruntime::MemoryInfoPerTensor>::iterator begin() { return map_.begin(); }
+  std::unordered_map<onnxruntime::OrtValueIndex, onnxruntime::MemoryInfoPerTensor>::const_iterator begin() const { return map_.begin(); }
+  std::unordered_map<onnxruntime::OrtValueIndex, onnxruntime::MemoryInfoPerTensor>::iterator end() { return map_.end(); }
+  std::unordered_map<onnxruntime::OrtValueIndex, onnxruntime::MemoryInfoPerTensor>::const_iterator end() const { return map_.end(); }
+  onnxruntime::MemoryInfoPerTensor& operator[](const OrtValueIndex& k) { return map_[k]; }
```

Comment:
>map_[k] [](start = 80, length = 7)

Are duplicates important? Especially with different values.

Reply:
Do you mean using vector instead of map? It will be looked up often, that's why I am using map.